docs/contributing.rst

```diff
@@ -158,35 +158,75 @@ Any extra texts (print statements and so on) should be removed.
 
      make test-3.10-no-extensions
 
-Tests coverage
---------------
+Code coverage
+=============
 
-We are trying hard to have good test coverage; please don't make it worse.
+We use *codecov.io* as an indispensable tool for analyzing our coverage
+results. Visit https://codecov.io/gh/aio-libs/aiohttp to see coverage
+reports for the master branch, history, pull requests etc.
 
-Use:
+We'll use an example from a real PR to demonstrate how we use this.
+Once the tests run in a PR, you'll see a comment posted by *codecov*.
+The most important thing to check here is whether there are any new
+missed or partial lines in the report:
 
-.. code-block:: shell
+.. image:: _static/img/contributing-cov-comment.svg
+
+Here, the PR has introduced 1 miss and 2 partials. Now we
+click the link in the comment header to open the full report:
+
+.. image:: _static/img/contributing-cov-header.svg
+   :alt: Codecov report
+
+Now, if we look through the diff under 'Files changed' we find one of
+our partials:
 
-   $ make cov-dev
+.. image:: _static/img/contributing-cov-partial.svg
+   :alt: A while loop with partial coverage.
 
-to run test suite and collect coverage information. Once the command
-has finished check your coverage at the file that appears in the last
-line of the output:
-``open file:///.../aiohttp/htmlcov/index.html``
+In this case, the while loop is never skipped in our tests. This is
+probably not worth writing a test for (and may be a situation that is
+impossible to trigger anyway), so we leave this alone.
 
-Please go to the link and make sure that your code change is covered.
+We're still missing a partial and a miss, so we switch to the
+'Indirect changes' tab and take a look through the diff there. This
+time we find the remaining 2 lines:
 
+.. image:: _static/img/contributing-cov-miss.svg
+   :alt: An if statement that isn't covered anymore.
 
-The project uses *codecov.io* for storing coverage results. Visit
-https://codecov.io/gh/aio-libs/aiohttp for looking on coverage of
-master branch, history, pull requests etc.
+After reviewing the PR, we find that this code is no longer needed as
+the changes mean that this method will never be called under those
+conditions. Thanks to this report, we were able to remove some
+redundant code from a performance-critical part of our codebase (this
+check would have been run, probably multiple times, for every single
+incoming request).
+
+.. tip::
+   Sometimes the diff on *codecov.io* doesn't make sense. This is usually
+   caused by the branch being out of sync with master. Try merging
```

Comment:
@Dreamsorcerer this isn't exactly the case I meant. What I'm talking about is GHA attempting to upload coverage and getting HTTP 500 from Codecov API in some or all of the jobs. Sometimes, restarting those jobs doesn't help and it just keeps erroring out for hours.
This means that part of the coverage uploads happened from some envs, but not all. And depending on which ones failed, the coverage may show up lower than it actually is.

To make things worse, by default, the GHA step doing the upload succeeds silently. And the only way to discover if some jobs didn't actually upload coverage is to click through each CI job, expand that step and read the log (it'll either have a URL of the upload for the current commit/PR or some error message at the end of the console output).

Reply:
> What I'm talking about is GHA attempting to upload coverage and getting HTTP 500 from Codecov API in some or all of the jobs.

Oh, I think I enabled the option to fail CI if an error occurs in upload (unless it was a different repo), so we shouldn't get that happening anymore.

Reply:
Different repo, probably. But still, I'm not sure if that option is what we want — it effectively marks entire jobs as failing. And whenever Codecov has platform issues, that means fully blocking merges for hours if not days.

Reply:
Problem is that we've had several repos which have been failing coverage uploads for a long time, due to configuration issues. I've personally not seen codecov fail on upload as an issue on their end yet.